### DISCUSSION TITLE
Evaluating Medley compatibility with Practical Common Lisp code: Chapters 30 and 31

### DISCUSSION THREAD

**Discussion by pamoroso** [Interlisp/medley#1650] [General]
## Chapters 30 and 31

This is the continuation of the work on issue #609. To evaluate the compatibility of Medley Common Lisp with ANSI Common Lisp I'm testing on Medley the code of Peter Seibel's book [Practical Common Lisp](https://gigamonkeys.com/book).

### Code

Chapters [30](https://gigamonkeys.com/book/practical-an-html-generation-library-the-interpreter) and [31](https://gigamonkeys.com/book/practical-an-html-generation-library-the-compiler) describe an HTML generation library. 

Since this is a more substantial project than those in earlier chapters I'm preparing to test the code in two steps. First, from Medley I manually load in the proper order the source files that come with the book. These files depend from a macro library described in [chapter 8](https://gigamonkeys.com/book/macros-defining-your-own) which I also load.

Once I manage to load the files correctly I will evaluated at a XCL Exec the example expressions of the chapter, with the addition of just a couple of missing calls.

The source files require the following modifications to the package definition and access code:

* substituting strings for keywords as package designators and exported symbols (otherwise I get occasional errors I'm trying to reproduce)
* commenting out the definition of `with-gensyms` which XCL already has
* replacing the `:use` lists of package definitions to include `LISP`, `XCL`, and `CLOS` (because system packages are slightly different on Medley; and because I'm using the Medley Online 19-Mar-2024 02:06:49 build in which some packages are partially broken as a side effect of the initial port of Sacla `LOOP`, so `CLOS` is required to avoid errors such as `Undefined car of form DEFCLASS`)

Here are the source files, renamed from `.lisp` to `.txt` as GitHub requires:

* [pclchap8-packages.txt](https://github.com/Interlisp/medley/files/14893515/pclchap8-packages.txt)
* [pclchap8.txt](https://github.com/Interlisp/medley/files/14893517/pclchap8.txt)
* [pclchap31-packages.txt](https://github.com/Interlisp/medley/files/14893514/pclchap31-packages.txt)
* [pclchap31-html.txt](https://github.com/Interlisp/medley/files/14893516/pclchap31-html.txt)

### File loading issue

I hit a snag with this setup as I'm unable to proceed due to a file loading issue. From a XCL Exec I load the source files in this order:

1. pclchap8-packages.lisp
2. pclchap8.lisp
3. pclchap31-packages.lisp
4. pclchap31-html.lisp

As soon as I load pclchap31-html.lisp I'm immediately logged out of Medley Online and the session ends.

**Comment by pamoroso** [comment]
Since loading pclchap31-html.lisp causes the termination of the Medley Online session, I used the bisection method to narrow down where the issue occurs. Instead of loading pclchap31-html.lisp after the files it depends on, I proceeded by loading a modified copy of pclchap31-html.lisp containing the code up to a line indicated by the bisection method.

The result is the issue lies between lines 316 and 320 (inclusive) of pclchap31-html.lisp. That portion of the file contains only this code:

```lisp
(defun parse-cons-form (sexp)
  (if (consp (first sexp))
    (parse-explicit-attributes-sexp sexp)
    (parse-implicit-attributes-sexp sexp)))
```

Nothing jumps at me but that code might just trigger an issue located elsehere.

**Comment by masinter** [comment]
[pclchap8-31.zip](https://github.com/Interlisp/medley/files/14901053/pclchap8-31.zip)

i got further with some problems I suspect in CLOS. The crash is a stack overflow that needs more than the stack reserve tobreak and needs RAID to hard reset, but isn't readily availabe online.  I suspect CLOS because the recursion involved IL:\INTERPRETER, IL:\REMOVE-DECLS etc. 
try compiling first and then loading the compiled files. see attached zip.
![image](https://github.com/Interlisp/medley/assets/1116587/2a312cbf-4cd1-4ad6-af8e-249b18efac2a)
![image](https://github.com/Interlisp/medley/assets/1116587/b0a2c1d1-9483-416b-8b8e-741bef810b86)

**Comment by pamoroso** [comment]
Now I can load the compiled files correctly. But when I evaluate test expressions I get undefined function errors like this:

```lisp
2/12> (in-package :com.gigamonkeys.html)
#<Package COM.GIGAMONKEYS.HTML>
2/13> 

2/13> *package*
#<Package COM.GIGAMONKEYS.HTML>
2/14> (apropos "PROCESS-SEXP-HTML")
PROCESS-SEXP-HTML
2/15> (html (:p "foo"))
In IL:OLDFAULT1:
PROCESS-SEXP-HTML is an undefined function.

2/16: ^

2/17>
```

`process-sexp-html` is defined in pclchap31-html.lisp and is internal to package `com.gigamonkeys.html`, but `in-package` should take care of that.

**Comment by pamoroso** [comment]
Never mind, the code of these chapters includes several calls to `write-sequence` which Medley Common Lisp doesn't support.
